packages/cli/src/evaluation.ee/test-runner/test-runner.service.ee.ts

```diff
@@ -246,151 +247,151 @@ export class TestRunnerService {
+			// Fetch only ids to reduce the data transfer.
+			const pastExecutions: ReadonlyArray<Pick<ExecutionEntity, 'id'>> =
+				await this.executionRepository
+					.createQueryBuilder('execution')
+					.select('execution.id')
+					.leftJoin('execution.annotation', 'annotation')
+					.leftJoin('annotation.tags', 'annotationTag')
+					.where('annotationTag.id = :tagId', { tagId: test.annotationTagId })
+					.andWhere('execution.workflowId = :workflowId', { workflowId: test.workflowId })
+					.getMany();
+
+			this.logger.debug('Found past executions', { count: pastExecutions.length });
+
+			// Get the metrics to collect from the evaluation workflow
+			const testMetricNames = await this.getTestMetricNames(test.id);
+
+			// 2. Run over all the test cases
+			await this.testRunRepository.markAsRunning(testRun.id, pastExecutions.length);
+
+			// Object to collect the results of the evaluation workflow executions
+			const metrics = new EvaluationMetrics(testMetricNames);
+
+			for (const { id: pastExecutionId } of pastExecutions) {
+				if (abortSignal.aborted) {
+					this.logger.debug('Test run was cancelled', {
+						testId: test.id,
+						stoppedOn: pastExecutionId,
+					});
+					break;
 				}
 
-				// Collect the results of the test case execution
-				const testCaseRunData = testCaseExecution.data.resultData.runData;
-
-				// Get the original runData from the test case execution data
-				const originalRunData = executionData.resultData.runData;
-
-				// Run the evaluation workflow with the original and new run data
-				const evalExecution = await this.runTestCaseEvaluation(
-					evaluationWorkflow,
-					originalRunData,
-					testCaseRunData,
-					testRun.id,
-				);
-				assert(evalExecution);
-
-				this.logger.debug('Evaluation execution finished', { pastExecutionId });
-
-				metrics.addResults(this.extractEvaluationResult(evalExecution));
-
-				if (evalExecution.data.resultData.error) {
+				this.logger.debug('Running test case', { pastExecutionId });
+
+				try {
+					// Fetch past execution with data
+					const pastExecution = await this.executionRepository.findOne({
+						where: { id: pastExecutionId },
+						relations: ['executionData', 'metadata'],
+					});
+					assert(pastExecution, 'Execution not found');
+
+					const executionData = parse(pastExecution.executionData.data) as IRunExecutionData;
+
+					// Run the test case and wait for it to finish
+					const testCaseExecution = await this.runTestCase(
+						workflow,
+						executionData,
+						pastExecution.executionData.workflowData,
+						test.mockedNodes,
+						user.id,
+						abortSignal,
+					);
+
+					this.logger.debug('Test case execution finished', { pastExecutionId });
+
+					// In case of a permission check issue, the test case execution will be undefined.
+					// Skip them, increment the failed count and continue with the next test case
+					if (!testCaseExecution) {
+						await this.testRunRepository.incrementFailed(testRun.id);
```

Comment:
Do we have plans on showing failure reasons somewhere? I'd imagine this might be pretty difficult to debug

Reply:
Yes, we plan to introduce one more entity which will be linking test runs and executions, and store errors there. This will allow us to show failure reasons on a test case level.